toolchain/check/facet_type.cpp

```diff
@@ -738,102 +575,87 @@ auto ResolveFacetTypeRewriteConstraints(
 
     rewrite_values.InsertNotRewritten(context, *lhs_access, constraint.rhs_id);
   }
+
   for (auto& constraint : rewrites) {
-    if (constraint.lhs_id == SemIR::ErrorInst::InstId ||
-        constraint.rhs_id == SemIR::ErrorInst::InstId) {
-      continue;
-    }
     auto lhs_access =
         context.insts().TryGetAs<SemIR::ImplWitnessAccess>(constraint.lhs_id);
     if (!lhs_access) {
       continue;
     }
 
-    rewrite_values.SetBeingRewritten(context, *lhs_access);
+    auto* lhs_rewrite_value = rewrite_values.FindRef(context, *lhs_access);
+    // Every LHS was added with InsertNotRewritten above.
+    CARBON_CHECK(lhs_rewrite_value);
+    rewrite_values.SetBeingRewritten(*lhs_rewrite_value);
 
     auto replace_witness_callbacks =
         SubstImplWitnessAccessCallbacks(&context, loc_id, &rewrite_values);
-    auto subst_inst_id =
+    auto rhs_subst_inst_id =
         SubstInst(context, constraint.rhs_id, replace_witness_callbacks);
-    constraint.rhs_id = subst_inst_id;
-    if (constraint.rhs_id == SemIR::ErrorInst::InstId) {
+    if (rhs_subst_inst_id == SemIR::ErrorInst::InstId) {
+      return false;
+    }
+
+    if (lhs_rewrite_value->state == AccessRewriteValues::FullyRewritten &&
+        !CompareFacetTypeConstraintValues(context, lhs_rewrite_value->inst_id,
+                                          rhs_subst_inst_id)) {
+      if (lhs_rewrite_value->inst_id != SemIR::ErrorInst::InstId) {
+        CARBON_DIAGNOSTIC(AssociatedConstantWithDifferentValues, Error,
+                          "associated constant {0} given two different "
+                          "values {1} and {2}",
+                          InstIdAsConstant, InstIdAsConstant, InstIdAsConstant);
+        // Use inst id ordering as a simple proxy for source ordering, to
+        // try to name the values in the same order they appear in the facet
+        // type.
+        auto source_order1 =
+            lhs_rewrite_value->inst_id.index < rhs_subst_inst_id.index
+                ? lhs_rewrite_value->inst_id
+                : rhs_subst_inst_id;
+        auto source_order2 =
+            lhs_rewrite_value->inst_id.index >= rhs_subst_inst_id.index
+                ? lhs_rewrite_value->inst_id
+                : rhs_subst_inst_id;
+        // TODO: It would be nice to note the places where the values are
+        // assigned but rewrite constraint instructions are from canonical
+        // constant values, and have no locations. We'd need to store a
+        // location along with them in the rewrite constraints.
+        context.emitter().Emit(loc_id, AssociatedConstantWithDifferentValues,
+                               constraint.lhs_id, source_order1, source_order2);
+      }
       return false;
     }
 
-    rewrite_values.SetFullyRewritten(context, *lhs_access, subst_inst_id);
+    rewrite_values.SetFullyRewritten(context, *lhs_rewrite_value,
+                                     rhs_subst_inst_id);
```

Comment:
Optional: We could make this an `else` for the `state == FullyRewritten` check above; if so, I think it could simply check that the previous state was `BeingRewritten`, which might be slightly simpler than having it also check the `inst_id` hasn't changed.